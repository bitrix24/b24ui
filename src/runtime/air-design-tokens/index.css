@import './001_b24_global.css';
@import './003_b24_context-content-light.css';
@import './004_b24_context-content-dark.css';
@import './005_b24_context-edge-light.css';
@import './006_b24_context-edge-dark.css';
@import './007_b24_global.css';
@import './008_ui_global.css';
@import './009_b24_tools.css';
@import './components/scrollbar.css';
@import './components/navigation-menu.css';
@import './components/badge-counter.css';
