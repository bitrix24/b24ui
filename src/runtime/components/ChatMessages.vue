<script lang="ts">
import type { AppConfig } from '@nuxt/schema'
import type { UIMessage, ChatStatus } from 'ai'
import theme from '#build/b24ui/chat-messages'
import type { ButtonProps, ChatMessageProps, ChatMessageSlots, IconComponent } from '../types'
import type { ComponentConfig } from '../types/tv'

type ChatMessages = ComponentConfig<typeof theme, AppConfig, 'chatMessages'>

export interface ChatMessagesProps {
  messages?: UIMessage[]
  status?: ChatStatus
  /**
   * Whether to automatically scroll to the bottom when a message is streaming.
   * @defaultValue false
   */
  shouldAutoScroll?: boolean
  /**
   * Whether to scroll to the bottom on mounted.
   * @defaultValue true
   */
  shouldScrollToBottom?: boolean
  /**
   * Display an auto scroll button.
   * `{ size: 'md', color: 'air-secondary-no-accent' }`{lang="ts-type"}
   * @defaultValue true
   */
  autoScroll?: boolean | Partial<ButtonProps>
  /**
   * The icon displayed in the auto scroll button.
   * @defaultValue icons.arrowDown
   * @IconComponent
   */
  autoScrollIcon?: IconComponent
  /**
   * The `user` messages props.
   * `{ side: 'right', variant: 'message' }`{lang="ts-type"}
   */
  user?: Pick<ChatMessageProps, 'icon' | 'avatar' | 'variant' | 'side' | 'actions' | 'b24ui'>
  /**
   * The `assistant` messages props.
   * `{ side: 'left', variant: 'message' }`{lang="ts-type"}
   */
  assistant?: Pick<ChatMessageProps, 'icon' | 'avatar' | 'variant' | 'side' | 'actions' | 'b24ui'>
  /**
   * Render the messages in a compact style.
   * This is done automatically when used inside a `UChatPalette`{lang="ts-type"}.
   * @defaultValue false
   */
  compact?: boolean
  /**
   * The spacing offset for the last message in px. Can be useful when the prompt is sticky for example.
   * @defaultValue 0
   */
  spacingOffset?: number
  class?: any
  b24ui?: ChatMessages['slots']
}

type ExtendSlotWithVersion<K extends keyof ChatMessageSlots>
  = ChatMessageSlots[K] extends (props: infer P) => any
    ? (props: P & { message: UIMessage }) => any
    : ChatMessageSlots[K]

export type ChatMessagesSlots = {
  [K in keyof ChatMessageSlots]: ExtendSlotWithVersion<K>
} & {
  default(props?: {}): any
  indicator(props: { b24ui: ChatMessages['b24ui'] }): any
  viewport(props: { b24ui: ChatMessages['b24ui'], onClick: () => void }): any
}
</script>

<script setup lang="ts">
import type { ComponentPublicInstance } from 'vue'
import { ref, computed, watch, nextTick, toRef, onMounted } from 'vue'
import { Presence } from 'reka-ui'
import { defu } from 'defu'
import { useElementBounding, useEventListener, watchThrottled } from '@vueuse/core'
import { useAppConfig } from '#imports'
import { omit } from '../utils'
import { tv } from '../utils/tv'
import icons from '../dictionary/icons'
import B24ChatMessage from './ChatMessage.vue'
import B24Button from './Button.vue'

const props = withDefaults(defineProps<ChatMessagesProps>(), {
  autoScroll: true,
  shouldAutoScroll: false,
  shouldScrollToBottom: true,
  spacingOffset: 0
})
const slots = defineSlots<ChatMessagesSlots>()

const getProxySlots = () => omit(slots, ['default', 'indicator', 'viewport'])

const appConfig = useAppConfig() as ChatMessages['AppConfig']

const userProps = toRef(() => defu(props.user, { side: 'right' as const, variant: 'message' as const }))
const assistantProps = toRef(() => defu(props.assistant, { side: 'left' as const, variant: 'message' as const }))

const b24ui = computed(() => tv({ extend: tv(theme), ...(appConfig.b24ui?.chatMessages || {}) })({
  compact: props.compact
}))

const el = ref<HTMLElement | null>(null)
const parent = ref<HTMLElement | null>(null)
const messagesRefs = ref(new Map<string, HTMLElement>())

const showAutoScroll = ref(false)
const lastMessageHeight = ref(0)
const lastMessageSubmitted = ref(false)
const lastScrollTop = ref(0)
const userScrolledUp = ref(false)

function registerMessageRef(id: string, element: ComponentPublicInstance | null) {
  const elInstance = element?.$el
  if (elInstance) {
    messagesRefs.value.set(id, elInstance)
  }
}

function scrollToMessage(id: string) {
  const element = messagesRefs.value.get(id)
  if (element) {
    element.scrollIntoView({ behavior: 'smooth', block: 'start' })
  }
}

function scrollToBottom(smooth: boolean = true) {
  if (!parent.value) {
    return
  }

  if (smooth) {
    parent.value.scrollTo({ top: parent.value.scrollHeight, behavior: 'smooth' })
  } else {
    parent.value.scrollTop = parent.value.scrollHeight
  }
}

watchThrottled([() => props.messages, () => props.status], ([_, status]) => {
  if (status !== 'streaming') {
    return
  }

  if (!props.shouldAutoScroll) {
    checkScrollPosition()
    return
  }

  // Scroll to bottom when message is streaming if `props.shouldAutoScroll` is true
  nextTick(() => {
    if (!parent.value || userScrolledUp.value) return

    if ((parent.value.scrollHeight - parent.value.scrollTop - parent.value.clientHeight) < 150) {
      scrollToBottom(false)
    }
  })
}, { deep: true, throttle: 50, leading: true })

watch(() => props.status, (status) => {
  if (status !== 'submitted') {
    return
  }

  const lastMessage = props.messages?.[props.messages.length - 1]
  if (!lastMessage || lastMessage.role !== 'user') {
    return
  }

  userScrolledUp.value = false

  nextTick(() => {
    lastMessageSubmitted.value = true

    updateLastMessageHeight()

    nextTick(() => {
      scrollToMessage(lastMessage.id)
    })
  })
})

function checkScrollPosition() {
  if (!parent.value) {
    return
  }

  const scrollPosition = parent.value.scrollTop + parent.value.clientHeight
  const scrollHeight = parent.value.scrollHeight
  const threshold = 100

  showAutoScroll.value = (scrollHeight - scrollPosition) >= threshold

  // Detect user scrolling up
  if (parent.value.scrollTop < lastScrollTop.value) {
    userScrolledUp.value = true
  } else if ((scrollHeight - scrollPosition) < threshold) {
    userScrolledUp.value = false
  }

  lastScrollTop.value = parent.value.scrollTop
}

function onAutoScrollClick() {
  userScrolledUp.value = false
  scrollToBottom()
}

function getScrollParent(node: HTMLElement | null): HTMLElement {
  if (!node) {
    return document.documentElement
  }

  const overflowRegex = /auto|scroll/

  let current: HTMLElement | null = node
  while (current && current !== document.body && current !== document.documentElement) {
    const style = window.getComputedStyle(current)
    if (overflowRegex.test(style.overflowY)) {
      return current
    }

    current = current.parentElement
  }

  return document.documentElement
}

function updateLastMessageHeight() {
  if (!el.value || !parent.value || !props.messages?.length || !lastMessageSubmitted.value) {
    return
  }

  const { height: parentHeight } = useElementBounding(parent.value)

  const lastMessage = props.messages.findLast(m => m.role === 'user')
  if (!lastMessage) {
    return
  }

  const lastMessageEl = messagesRefs.value.get(lastMessage.id)
  if (!lastMessageEl) {
    return
  }

  let spacingOffset = props.spacingOffset || 0
  const elComputedStyle = window.getComputedStyle(el.value)
  const parentComputedStyle = window.getComputedStyle(parent.value)

  spacingOffset += Number.parseFloat(elComputedStyle.rowGap) || Number.parseFloat(elComputedStyle.gap) || 0
  spacingOffset += Number.parseFloat(parentComputedStyle.paddingTop) || 0
  spacingOffset += Number.parseFloat(parentComputedStyle.paddingBottom) || 0

  lastMessageHeight.value = Math.max(parentHeight.value - lastMessageEl.offsetHeight - spacingOffset, 0)
}

onMounted(() => {
  // Find first scrollable parent element
  parent.value = getScrollParent(el.value)
  if (!parent.value) {
    return
  }

  lastScrollTop.value = parent.value.scrollTop

  // Wait for content to fully render (especially MDC components in ChatPalette)
  setTimeout(() => {
    if (props.shouldScrollToBottom) {
      // Scroll to bottom on mount without smooth animation when `props.shouldScrollToBottom` is true
      scrollToBottom(false)
    } else {
      checkScrollPosition()
    }
  }, 100)

  // Add event listener to check scroll position to show the auto scroll button
  useEventListener(parent, 'scroll', checkScrollPosition)

  // Add event listener to update the last message height when the window is resized
  useEventListener(window, 'resize', () => nextTick(updateLastMessageHeight))
})
</script>

<template>
  <div
    ref="el"
    :data-status="status"
    :class="b24ui.root({ class: [props.b24ui?.root, props.class] })"
    :style="{ '--last-message-height': `${lastMessageHeight}px` }"
  >
    <slot>
      <B24ChatMessage
        v-for="message in messages"
        :key="message.id"
        v-bind="{ ...message, ...(message.role === 'user' ? userProps : assistantProps) }"
        :ref="(el) => registerMessageRef(message.id, el as ComponentPublicInstance)"
        :compact="compact"
      >
        <template v-for="(_, name) in getProxySlots()" #[name]="slotData">
          <slot :name="name" v-bind="(slotData as any)" :message="message" />
        </template>
      </B24ChatMessage>
    </slot>

    <B24ChatMessage
      v-if="status === 'submitted'"
      id="indicator"
      role="assistant"
      v-bind="{ ...assistantProps, actions: undefined, parts: [] }"
      :compact="compact"
    >
      <template #content>
        <slot name="indicator" :b24ui="b24ui">
          <div :class="b24ui.indicator({ class: props.b24ui?.indicator })">
            <span />
            <span />
            <span />
          </div>
        </slot>
      </template>
    </B24ChatMessage>

    <Presence :present="showAutoScroll">
      <div :data-state="showAutoScroll ? 'open' : 'closed'" :class="b24ui.viewport({ class: props.b24ui?.viewport })">
        <slot name="viewport" :b24ui="b24ui" :on-click="onAutoScrollClick">
          <B24Button
            v-if="autoScroll"
            :icon="autoScrollIcon || icons.arrowDown"
            color="air-secondary-no-accent"
            v-bind="(typeof autoScroll === 'object' ? autoScroll as Partial<ButtonProps> : {})"
            :class="b24ui.autoScroll({ class: props.b24ui?.autoScroll })"
            @click="onAutoScrollClick"
          />
        </slot>
      </div>
    </Presence>
  </div>
</template>
