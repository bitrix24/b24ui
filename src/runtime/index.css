@plugin '@bitrix24/b24style';
@import '#build/b24ui.css';
@import './keyframes.css';
@import './air-design-tokens/index.css';

@layer base {
  body,
  .sidebar-layout.--inner {
    /* @see air-design-tokens/???_b24_context*.css */
    --air-theme-background:
      var(--air-theme-bg-image, none)
      var(--air-theme-bg-position, 0 0) / var(--air-theme-bg-size, auto)
      var(--air-theme-bg-repeat, repeat)
      var(--air-theme-bg-attachment, scroll)
      var(--air-theme-bg-color, transparent)
    ;

    --air-theme-background-blurred:
      var(--air-theme-bg-image-blurred, none)
      var(--air-theme-bg-position, 0 0) / var(--air-theme-bg-size, auto)
      var(--air-theme-bg-repeat, repeat)
      var(--air-theme-bg-attachment, scroll)
    ;
  }

  /* @see src/theme/sidebar-layout.ts */
  body {
    background: var(--air-theme-background);
  }

  .sidebar-layout.--inner {
    background: var(--air-theme-background);
  }

  body {
    overflow-y: scroll!important;
    padding-right: 0!important;
    scrollbar-gutter: stable;
  }

  .--app[data-state="loading"] {
    overflow: hidden;
  }
}
