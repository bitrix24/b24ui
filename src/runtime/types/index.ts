export * from '../components/App.vue'
export * from '../components/Avatar.vue'
export * from '../components/AvatarGroup.vue'
export * from '../components/Button.vue'
export * from '../components/Checkbox.vue'
export * from '../components/Chip.vue'
export * from '../components/Container.vue'
export * from '../components/FormField.vue'
export * from '../components/Input.vue'
export * from '../components/Kbd.vue'
export * from '../components/Link.vue'
export * from '../components/Progress.vue'
export * from '../components/RadioGroup.vue'
export * from '../components/Separator.vue'
export * from '../components/Skeleton.vue'
export * from '../components/content/SkeletonTaskAdd.vue'
export * from '../components/Switch.vue'
export * from '../components/Textarea.vue'
export * from '../components/Tooltip.vue'
export * from './form'
export * from './locale'
