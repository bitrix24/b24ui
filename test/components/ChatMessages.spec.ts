import { describe, it, expect } from 'vitest'
import { axe } from 'vitest-axe'
import { mountSuspended } from '@nuxt/test-utils/runtime'
import ChatMessages from '../../src/runtime/components/ChatMessages.vue'
import type { ChatMessagesProps, ChatMessagesSlots } from '../../src/runtime/components/ChatMessages.vue'
import ComponentRender from '../component-render'
import Search2Icon from '@bitrix24/b24icons-vue/main/Search2Icon'

describe('ChatMessages', () => {
  const statuses = ['ready', 'submitted', 'streaming', 'error'] as any

  const props = {
    messages: [{
      id: '6045235a-a435-46b8-989d-2df38ca2eb47',
      role: 'user' as const,
      parts: [{ type: 'text' as const, text: 'Hello, how are you?' }]
    }, {
      id: '6045235a-a435-46b8-989d-2df38ca2eb47',
      role: 'assistant' as const,
      parts: [{ type: 'text' as const, text: 'I am fine, thank you!' }]
    }]
  }

  it.each([
    // Props
    ['with messages', { props }],
    ['with user', { props: { ...props, user: { avatar: { src: 'https://github.com/bitrix24.png' }, variant: 'soft', side: 'right' } } }],
    ['with assistant', { props: { ...props, assistant: { avatar: { icon: Search2Icon } } } }],
    ...statuses.map((status: string) => [`with status ${status}`, { props: { ...props, status } }]),
    ['with compact', { props: { ...props, compact: true } }],
    ['with class', { props: { ...props, class: 'max-w-3xl' } }],
    ['with b24ui', { props: { ...props, b24ui: { autoScroll: 'bottom-0' } } }],
    // Slots
    ['with default slot', { props, slots: { default: () => 'Default slot' } }],
    ['with indicator slot', { props: { ...props, status: 'submitted' }, slots: { indicator: () => 'Indicator slot' } }],
    ['with viewport slot', { props, slots: { viewport: () => 'Viewport slot' } }],
    ['with content slot', { props, slots: { content: () => 'Content slot' } }]
  ])('renders %s correctly', async (nameOrHtml: string, options: { props?: ChatMessagesProps, slots?: Partial<ChatMessagesSlots> }) => {
    const html = await ComponentRender(nameOrHtml, options, ChatMessages)
    expect(html).toMatchSnapshot()
  })

  it('passes accessibility tests', async () => {
    const wrapper = await mountSuspended(ChatMessages, {
      props
    })

    expect(await axe(wrapper.element)).toHaveNoViolations()
  })
})
