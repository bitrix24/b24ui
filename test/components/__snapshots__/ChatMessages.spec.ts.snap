// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChatMessages > renders with assistant correctly 1`] = `
"<div class="w-full flex flex-col gap-1 flex-1 px-2.5 [&amp;>article]:last-of-type:min-h-(--last-message-height)" style="--last-message-height: 0px;">
  <article data-role="user" class="group/message relative w-full scroll-mt-4 sm:scroll-mt-6">
    <div class="relative flex items-start ltr:justify-end ms-auto max-w-[75%] gap-3 pb-8">
      <!--v-if-->
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0 bg-inverted text-inverted px-4 py-3 rounded-lg min-h-12">Hello, how are you?</div>
      <!--v-if-->
    </div>
  </article>
  <article data-role="assistant" class="group/message relative w-full scroll-mt-4 sm:scroll-mt-6">
    <div class="relative flex items-start rtl:justify-end gap-3 pb-8">
      <div class="inline-flex items-center justify-center min-h-6 mt-2"><span class="air-secondary-accent inline-flex items-center justify-center select-none rounded-full align-middle bg-(--ui-color-base-8) ring ring-(--ui-color-base-7) size-[32px] text-(length:--ui-font-size-sm)/(--ui-font-line-height-reset) shrink-0"><svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" aria-hidden="true" data-slot="icon" class="text-(--ui-color-design-plain-content-icon-secondary) shrink-0 size-[28px]"><path fill="currentColor" fill-rule="evenodd" d="M14.05 15.886a6.37 6.37 0 1 1 1.836-1.837l3.311 3.311a1 1 0 0 1 0 1.414l-.422.423a1 1 0 0 1-1.414 0zm1.058-5.328a4.55 4.55 0 1 1-9.098 0 4.55 4.55 0 0 1 9.099 0" clip-rule="evenodd"></path></svg></span></div>
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0 bg-inverted text-inverted px-4 py-3 rounded-lg min-h-12 w-full">I am fine, thank you!</div>
      <!--v-if-->
    </div>
  </article>
  <!--v-if-->
  <!---->
</div>"
`;

exports[`ChatMessages > renders with b24ui correctly 1`] = `
"<div class="w-full flex flex-col gap-1 flex-1 px-2.5 [&amp;>article]:last-of-type:min-h-(--last-message-height)" style="--last-message-height: 0px;">
  <article data-role="user" class="group/message relative w-full scroll-mt-4 sm:scroll-mt-6">
    <div class="relative flex items-start ltr:justify-end ms-auto max-w-[75%] gap-3 pb-8">
      <!--v-if-->
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0 bg-inverted text-inverted px-4 py-3 rounded-lg min-h-12">Hello, how are you?</div>
      <!--v-if-->
    </div>
  </article>
  <article data-role="assistant" class="group/message relative w-full scroll-mt-4 sm:scroll-mt-6">
    <div class="relative flex items-start rtl:justify-end gap-3 pb-8">
      <!--v-if-->
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0 bg-inverted text-inverted px-4 py-3 rounded-lg min-h-12 w-full">I am fine, thank you!</div>
      <!--v-if-->
    </div>
  </article>
  <!--v-if-->
  <!---->
</div>"
`;

exports[`ChatMessages > renders with class correctly 1`] = `
"<div class="w-full flex flex-col gap-1 flex-1 px-2.5 [&amp;>article]:last-of-type:min-h-(--last-message-height) max-w-3xl" style="--last-message-height: 0px;">
  <article data-role="user" class="group/message relative w-full scroll-mt-4 sm:scroll-mt-6">
    <div class="relative flex items-start ltr:justify-end ms-auto max-w-[75%] gap-3 pb-8">
      <!--v-if-->
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0 bg-inverted text-inverted px-4 py-3 rounded-lg min-h-12">Hello, how are you?</div>
      <!--v-if-->
    </div>
  </article>
  <article data-role="assistant" class="group/message relative w-full scroll-mt-4 sm:scroll-mt-6">
    <div class="relative flex items-start rtl:justify-end gap-3 pb-8">
      <!--v-if-->
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0 bg-inverted text-inverted px-4 py-3 rounded-lg min-h-12 w-full">I am fine, thank you!</div>
      <!--v-if-->
    </div>
  </article>
  <!--v-if-->
  <!---->
</div>"
`;

exports[`ChatMessages > renders with compact correctly 1`] = `
"<div class="w-full flex flex-col gap-1 flex-1 px-2.5 [&amp;>article]:last-of-type:min-h-(--last-message-height)" style="--last-message-height: 0px;">
  <article data-role="user" class="group/message relative w-full scroll-mt-3">
    <div class="relative flex items-start ltr:justify-end ms-auto max-w-[75%] gap-1.5 pb-3">
      <!--v-if-->
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0 bg-inverted text-inverted px-2 py-1 rounded-lg min-h-8">Hello, how are you?</div>
      <!--v-if-->
    </div>
  </article>
  <article data-role="assistant" class="group/message relative w-full scroll-mt-3">
    <div class="relative flex items-start rtl:justify-end gap-1.5 pb-3">
      <!--v-if-->
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0 bg-inverted text-inverted px-2 py-1 rounded-lg min-h-8 w-full">I am fine, thank you!</div>
      <!--v-if-->
    </div>
  </article>
  <!--v-if-->
  <!---->
</div>"
`;

exports[`ChatMessages > renders with content slot correctly 1`] = `
"<div class="w-full flex flex-col gap-1 flex-1 px-2.5 [&amp;>article]:last-of-type:min-h-(--last-message-height)" style="--last-message-height: 0px;">
  <article data-role="user" class="group/message relative w-full scroll-mt-4 sm:scroll-mt-6">
    <div class="relative flex items-start ltr:justify-end ms-auto max-w-[75%] gap-3 pb-8">
      <!--v-if-->
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0 bg-inverted text-inverted px-4 py-3 rounded-lg min-h-12">Content slot</div>
      <!--v-if-->
    </div>
  </article>
  <article data-role="assistant" class="group/message relative w-full scroll-mt-4 sm:scroll-mt-6">
    <div class="relative flex items-start rtl:justify-end gap-3 pb-8">
      <!--v-if-->
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0 bg-inverted text-inverted px-4 py-3 rounded-lg min-h-12 w-full">Content slot</div>
      <!--v-if-->
    </div>
  </article>
  <!--v-if-->
  <!---->
</div>"
`;

exports[`ChatMessages > renders with default slot correctly 1`] = `
"<div class="w-full flex flex-col gap-1 flex-1 px-2.5 [&amp;>article]:last-of-type:min-h-(--last-message-height)" style="--last-message-height: 0px;">Default slot
  <!--v-if-->
  <!---->
</div>"
`;

exports[`ChatMessages > renders with indicator slot correctly 1`] = `
"<div data-status="submitted" class="w-full flex flex-col gap-1 flex-1 px-2.5 [&amp;>article]:last-of-type:min-h-(--last-message-height)" style="--last-message-height: 0px;">
  <article data-role="user" class="group/message relative w-full scroll-mt-4 sm:scroll-mt-6">
    <div class="relative flex items-start ltr:justify-end ms-auto max-w-[75%] gap-3 pb-8">
      <!--v-if-->
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0 bg-inverted text-inverted px-4 py-3 rounded-lg min-h-12">Hello, how are you?</div>
      <!--v-if-->
    </div>
  </article>
  <article data-role="assistant" class="group/message relative w-full scroll-mt-4 sm:scroll-mt-6">
    <div class="relative flex items-start rtl:justify-end gap-3 pb-8">
      <!--v-if-->
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0 bg-inverted text-inverted px-4 py-3 rounded-lg min-h-12 w-full">I am fine, thank you!</div>
      <!--v-if-->
    </div>
  </article>
  <article data-role="assistant" class="group/message relative w-full scroll-mt-4 sm:scroll-mt-6">
    <div class="relative flex items-start rtl:justify-end gap-3 pb-8">
      <!--v-if-->
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0 bg-inverted text-inverted px-4 py-3 rounded-lg min-h-12 w-full">Indicator slot</div>
      <!--v-if-->
    </div>
  </article>
  <!---->
</div>"
`;

exports[`ChatMessages > renders with messages correctly 1`] = `
"<div class="w-full flex flex-col gap-1 flex-1 px-2.5 [&amp;>article]:last-of-type:min-h-(--last-message-height)" style="--last-message-height: 0px;">
  <article data-role="user" class="group/message relative w-full scroll-mt-4 sm:scroll-mt-6">
    <div class="relative flex items-start ltr:justify-end ms-auto max-w-[75%] gap-3 pb-8">
      <!--v-if-->
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0 bg-inverted text-inverted px-4 py-3 rounded-lg min-h-12">Hello, how are you?</div>
      <!--v-if-->
    </div>
  </article>
  <article data-role="assistant" class="group/message relative w-full scroll-mt-4 sm:scroll-mt-6">
    <div class="relative flex items-start rtl:justify-end gap-3 pb-8">
      <!--v-if-->
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0 bg-inverted text-inverted px-4 py-3 rounded-lg min-h-12 w-full">I am fine, thank you!</div>
      <!--v-if-->
    </div>
  </article>
  <!--v-if-->
  <!---->
</div>"
`;

exports[`ChatMessages > renders with status error correctly 1`] = `
"<div data-status="error" class="w-full flex flex-col gap-1 flex-1 px-2.5 [&amp;>article]:last-of-type:min-h-(--last-message-height)" style="--last-message-height: 0px;">
  <article data-role="user" class="group/message relative w-full scroll-mt-4 sm:scroll-mt-6">
    <div class="relative flex items-start ltr:justify-end ms-auto max-w-[75%] gap-3 pb-8">
      <!--v-if-->
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0 bg-inverted text-inverted px-4 py-3 rounded-lg min-h-12">Hello, how are you?</div>
      <!--v-if-->
    </div>
  </article>
  <article data-role="assistant" class="group/message relative w-full scroll-mt-4 sm:scroll-mt-6">
    <div class="relative flex items-start rtl:justify-end gap-3 pb-8">
      <!--v-if-->
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0 bg-inverted text-inverted px-4 py-3 rounded-lg min-h-12 w-full">I am fine, thank you!</div>
      <!--v-if-->
    </div>
  </article>
  <!--v-if-->
  <!---->
</div>"
`;

exports[`ChatMessages > renders with status ready correctly 1`] = `
"<div data-status="ready" class="w-full flex flex-col gap-1 flex-1 px-2.5 [&amp;>article]:last-of-type:min-h-(--last-message-height)" style="--last-message-height: 0px;">
  <article data-role="user" class="group/message relative w-full scroll-mt-4 sm:scroll-mt-6">
    <div class="relative flex items-start ltr:justify-end ms-auto max-w-[75%] gap-3 pb-8">
      <!--v-if-->
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0 bg-inverted text-inverted px-4 py-3 rounded-lg min-h-12">Hello, how are you?</div>
      <!--v-if-->
    </div>
  </article>
  <article data-role="assistant" class="group/message relative w-full scroll-mt-4 sm:scroll-mt-6">
    <div class="relative flex items-start rtl:justify-end gap-3 pb-8">
      <!--v-if-->
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0 bg-inverted text-inverted px-4 py-3 rounded-lg min-h-12 w-full">I am fine, thank you!</div>
      <!--v-if-->
    </div>
  </article>
  <!--v-if-->
  <!---->
</div>"
`;

exports[`ChatMessages > renders with status streaming correctly 1`] = `
"<div data-status="streaming" class="w-full flex flex-col gap-1 flex-1 px-2.5 [&amp;>article]:last-of-type:min-h-(--last-message-height)" style="--last-message-height: 0px;">
  <article data-role="user" class="group/message relative w-full scroll-mt-4 sm:scroll-mt-6">
    <div class="relative flex items-start ltr:justify-end ms-auto max-w-[75%] gap-3 pb-8">
      <!--v-if-->
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0 bg-inverted text-inverted px-4 py-3 rounded-lg min-h-12">Hello, how are you?</div>
      <!--v-if-->
    </div>
  </article>
  <article data-role="assistant" class="group/message relative w-full scroll-mt-4 sm:scroll-mt-6">
    <div class="relative flex items-start rtl:justify-end gap-3 pb-8">
      <!--v-if-->
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0 bg-inverted text-inverted px-4 py-3 rounded-lg min-h-12 w-full">I am fine, thank you!</div>
      <!--v-if-->
    </div>
  </article>
  <!--v-if-->
  <!---->
</div>"
`;

exports[`ChatMessages > renders with status submitted correctly 1`] = `
"<div data-status="submitted" class="w-full flex flex-col gap-1 flex-1 px-2.5 [&amp;>article]:last-of-type:min-h-(--last-message-height)" style="--last-message-height: 0px;">
  <article data-role="user" class="group/message relative w-full scroll-mt-4 sm:scroll-mt-6">
    <div class="relative flex items-start ltr:justify-end ms-auto max-w-[75%] gap-3 pb-8">
      <!--v-if-->
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0 bg-inverted text-inverted px-4 py-3 rounded-lg min-h-12">Hello, how are you?</div>
      <!--v-if-->
    </div>
  </article>
  <article data-role="assistant" class="group/message relative w-full scroll-mt-4 sm:scroll-mt-6">
    <div class="relative flex items-start rtl:justify-end gap-3 pb-8">
      <!--v-if-->
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0 bg-inverted text-inverted px-4 py-3 rounded-lg min-h-12 w-full">I am fine, thank you!</div>
      <!--v-if-->
    </div>
  </article>
  <article data-role="assistant" class="group/message relative w-full scroll-mt-4 sm:scroll-mt-6">
    <div class="relative flex items-start rtl:justify-end gap-3 pb-8">
      <!--v-if-->
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0 bg-inverted text-inverted px-4 py-3 rounded-lg min-h-12 w-full">
        <div class="h-6 flex items-center gap-1 py-3 *:size-2 *:rounded-full *:bg-(--ui-color-base-7) [&amp;>*:nth-child(1)]:animate-[bounce_1s_infinite] [&amp;>*:nth-child(2)]:animate-[bounce_1s_0.15s_infinite] [&amp;>*:nth-child(3)]:animate-[bounce_1s_0.3s_infinite]"><span></span><span></span><span></span></div>
      </div>
      <!--v-if-->
    </div>
  </article>
  <!---->
</div>"
`;

exports[`ChatMessages > renders with user correctly 1`] = `
"<div class="w-full flex flex-col gap-1 flex-1 px-2.5 [&amp;>article]:last-of-type:min-h-(--last-message-height)" style="--last-message-height: 0px;">
  <article data-role="user" class="group/message relative w-full scroll-mt-4 sm:scroll-mt-6">
    <div class="relative flex items-start ltr:justify-end ms-auto max-w-[75%] gap-3 pb-8">
      <div class="inline-flex items-center justify-center min-h-6"><span class="air-secondary-accent inline-flex items-center justify-center select-none rounded-full align-middle bg-(--ui-color-base-8) ring ring-(--ui-color-base-7) size-[32px] text-(length:--ui-font-size-sm)/(--ui-font-line-height-reset) shrink-0"><img src="https://github.com/bitrix24.png" width="32" height="32" class="h-full w-full rounded-[inherit] object-cover"></span></div>
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0">Hello, how are you?</div>
      <!--v-if-->
    </div>
  </article>
  <article data-role="assistant" class="group/message relative w-full scroll-mt-4 sm:scroll-mt-6">
    <div class="relative flex items-start rtl:justify-end gap-3 pb-8">
      <!--v-if-->
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0 bg-inverted text-inverted px-4 py-3 rounded-lg min-h-12 w-full">I am fine, thank you!</div>
      <!--v-if-->
    </div>
  </article>
  <!--v-if-->
  <!---->
</div>"
`;

exports[`ChatMessages > renders with viewport slot correctly 1`] = `
"<div class="w-full flex flex-col gap-1 flex-1 px-2.5 [&amp;>article]:last-of-type:min-h-(--last-message-height)" style="--last-message-height: 0px;">
  <article data-role="user" class="group/message relative w-full scroll-mt-4 sm:scroll-mt-6">
    <div class="relative flex items-start ltr:justify-end ms-auto max-w-[75%] gap-3 pb-8">
      <!--v-if-->
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0 bg-inverted text-inverted px-4 py-3 rounded-lg min-h-12">Hello, how are you?</div>
      <!--v-if-->
    </div>
  </article>
  <article data-role="assistant" class="group/message relative w-full scroll-mt-4 sm:scroll-mt-6">
    <div class="relative flex items-start rtl:justify-end gap-3 pb-8">
      <!--v-if-->
      <div class="relative text-pretty min-w-0 *:first:mt-0 *:last:mb-0 bg-inverted text-inverted px-4 py-3 rounded-lg min-h-12 w-full">I am fine, thank you!</div>
      <!--v-if-->
    </div>
  </article>
  <!--v-if-->
  <!---->
</div>"
`;
